# Data from Top200 SpotifyCharts


Proceso:
 - Sacar top200 de 2017, 2018, 2019 y 2010
 - De canciones duplicadas sumar los Streams y unir, quitar duplicados
 - Juntar Dataframes de años y añadir una nueva columna con el mismo.


## Preparación

In [6]:
# Help: https://github.com/kelvingakuo/fycharts#in
#! pip install fycharts

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime, timedelta
from datetime import date

#Fechas 2020
fecha_fin_2020 = '2020-05-13' # Empieza a levantarse restricciones de salida
fecha_inicial_2020 = '2020-03-14' #14 Marzo se decreta el estado de alarma, con 5753 casos y 136 fallecidos
fecha_fin2_2020 = datetime.strptime(fecha_fin_2020, '%Y-%m-%d')
fecha_inicial2_2020 = datetime.strptime(fecha_inicial_2020, '%Y-%m-%d')

#Fechas 2019
fecha_fin_2019 = '2019-05-13'
fecha_inicial_2019 = '2019-03-14'
fecha_fin2_2019 = datetime.strptime(fecha_fin_2019, '%Y-%m-%d')
fecha_inicial2_2019 = datetime.strptime(fecha_inicial_2019, '%Y-%m-%d')

print('Días totales seleccionados 2020:', fecha_fin2_2020 - fecha_inicial2_2020)
print('Días totales seleccionados 2019:', fecha_fin2_2019 - fecha_inicial2_2019)

Días totales seleccionados 2020: 60 days, 0:00:00
Días totales seleccionados 2019: 60 days, 0:00:00


In [7]:
# ¡IMPORTANTE! Ejecutar una única vez para descargar los datos, configurado para España
'''
# Descarga año 2020
from fycharts.SpotifyCharts import SpotifyCharts
import sqlalchemy

api = SpotifyCharts()
connector = sqlalchemy.create_engine("sqlite:///spotifycharts_2020.db", echo=False)
api.top200Daily(output_file = "top_200_daily_2020.csv", output_db = connector, webhook = "https://mywebhookssite.com/post/", 
                start = fecha_inicial_2020, end = fecha_fin_2020, region = ["es"])


# Descarga año 2019
from fycharts.SpotifyCharts import SpotifyCharts
import sqlalchemy

api = SpotifyCharts()
connector = sqlalchemy.create_engine("sqlite:///spotifycharts_2019.db", echo=False)
api.top200Daily(output_file = "top_200_daily_2019.csv", output_db = connector, webhook = "https://mywebhookssite.com/post/", 
                start = fecha_inicial_2019, end = fecha_fin_2019, region = ["es"])
'''

'\n# Descarga año 2020\nfrom fycharts.SpotifyCharts import SpotifyCharts\nimport sqlalchemy\n\napi = SpotifyCharts()\nconnector = sqlalchemy.create_engine("sqlite:///spotifycharts_2020.db", echo=False)\napi.top200Daily(output_file = "top_200_daily_2020.csv", output_db = connector, webhook = "https://mywebhookssite.com/post/", \n                start = fecha_inicial_2020, end = fecha_fin_2020, region = ["es"])\n\n\n# Descarga año 2019\nfrom fycharts.SpotifyCharts import SpotifyCharts\nimport sqlalchemy\n\napi = SpotifyCharts()\nconnector = sqlalchemy.create_engine("sqlite:///spotifycharts_2019.db", echo=False)\napi.top200Daily(output_file = "top_200_daily_2019.csv", output_db = connector, webhook = "https://mywebhookssite.com/post/", \n                start = fecha_inicial_2019, end = fecha_fin_2019, region = ["es"])\n'

## Importar la tabla

In [8]:
# Importamos la tabla con los datos y miramos a ver qué tal están
data_top200_2020= pd.read_csv("top_200_daily_2020.csv")
data_top200_2019= pd.read_csv("top_200_daily_2019.csv")
print(data_top200_2020.isnull().sum())
print('2020:', data_top200_2020.shape)
print('2019:',data_top200_2019.shape)
data_top200_2020.head()

Position      0
Track Name    0
Artist        0
Streams       0
date          0
region        0
spotify_id    0
dtype: int64
2020: (12200, 7)
2019: (12200, 7)


Position  Track Name          Artist  Streams        date region  \
0         1        Tusa         KAROL G   446086  2020-03-14     es   
1         2  La Difícil       Bad Bunny   399788  2020-03-14     es   
2         3      Tattoo  Rauw Alejandro   380503  2020-03-14     es   
3         4       Diosa     Myke Towers   336910  2020-03-14     es   
4         5        Rojo        J Balvin   327388  2020-03-14     es   

               spotify_id  
0  7k4t7uLgtOxPwTpFmtJNTY  
1  6NfrH0ANGmgBXyxgV2PeXt  
2  7na7Bk98usp84FaOJFPv3d  
3  3JHpk0DOTOzyh0777JFAky  
4  4uziEsK1yiqdauKVDPsmVG

In [9]:
# Importamos con las columnas que nos interesan
data_top200_2020 = pd.read_csv("top_200_daily_2020.csv", parse_dates= ["date"], usecols=['Position','Track Name', 'Artist', 'date', 'Streams', 'date', 'spotify_id'])
data_top200_2019 = pd.read_csv("top_200_daily_2019.csv", parse_dates= ["date"], usecols=['Position','Track Name', 'Artist', 'date', 'Streams', 'date', 'spotify_id'])
#Meto el año en nueva columnas columnas
data_top200_2020['year'] = pd.DatetimeIndex(data_top200_2020['date']).year
data_top200_2019['year'] = pd.DatetimeIndex(data_top200_2019['date']).year

#Concateno los datos de los dos años
data_top200 = pd.concat([data_top200_2020, data_top200_2019], axis=0,)

print(data_top200.dtypes)
print(data_top200.nunique()) # Ojo, debería haber el mismo número de Track Names y spotify ID!!
print(data_top200.shape)

Position               int64
Track Name            object
Artist                object
Streams                int64
date          datetime64[ns]
spotify_id            object
year                   int64
dtype: object
Position        200
Track Name      738
Artist          297
Streams       20942
date            122
spotify_id      807
year              2
dtype: int64
(24400, 7)


In [10]:
data_top200.loc[data_top200['spotify_id'] == '6NfrH0ANGmgBXyxgV2PeXt']

Position  Track Name     Artist  Streams       date  \
1             2  La Difícil  Bad Bunny   399788 2020-03-14   
201           2  La Difícil  Bad Bunny   372705 2020-03-15   
401           2  La Difícil  Bad Bunny   389384 2020-03-16   
601           2  La Difícil  Bad Bunny   390067 2020-03-17   
801           2  La Difícil  Bad Bunny   392738 2020-03-18   
...         ...         ...        ...      ...        ...   
11217        18  La Difícil  Bad Bunny   165292 2020-05-09   
11426        27  La Difícil  Bad Bunny   151313 2020-05-10   
11628        29  La Difícil  Bad Bunny   166240 2020-05-11   
11828        29  La Difícil  Bad Bunny   143671 2020-05-12   
12029        30  La Difícil  Bad Bunny   144698 2020-05-13   

                   spotify_id  year  
1      6NfrH0ANGmgBXyxgV2PeXt  2020  
201    6NfrH0ANGmgBXyxgV2PeXt  2020  
401    6NfrH0ANGmgBXyxgV2PeXt  2020  
601    6NfrH0ANGmgBXyxgV2PeXt  2020  
801    6NfrH0ANGmgBXyxgV2PeXt  2020  
...                       ...   ...  
11217  6NfrH0ANGmgBXyxgV2PeXt  2020  
11426  6NfrH0ANGmgBXyxgV2PeXt  2020  
11628  6NfrH0ANGmgBXyxgV2PeXt  2020  
11828  6NfrH0ANGmgBXyxgV2PeXt  2020  
12029  6NfrH0ANGmgBXyxgV2PeXt  2020  

[61 rows x 7 columns]

In [11]:
# Agrupar sumas y convertir en Dataframe
data_top200_sumastreams = data_top200.groupby('spotify_id')['Streams'].sum()
data_top200_sumastreams = pd.DataFrame({'spotify_id':data_top200_sumastreams.index, 'Streams':data_top200_sumastreams.values})
print(data_top200_sumastreams.shape)
data_top200_sumastreams.head()

(807, 2)


spotify_id  Streams
0  00ZBADBKZGwnzGIAA6U9Fb    44107
1  017PF4Q3l4DBUiWoXk4OWT  2814658
2  047WmwIeerHyIUstFAEz5A  3060737
3  04wvWMRKKxK9TGG4IPk32d    90156
4  059bcIhyc2SBwm6sw2AZzd  6447442

Tengo 24400 filas en canciones, de las cuales hay 738 canciones únicas (Dataframe 1, si hago un groupby para que me sume los Streams de las mismas canciones se hace sin problema y tengo un Dataframe 2

Pero ahora no quiero perder las demás columnas haciendo groupby, pero es la forma fácil de hacer las sumas de Streams. Pero entonces quisiera añadir en nueva columna en el dataframe 1 el dato de Streams totales de dicha canción. Lo ideal sería hacer un bucle que mirara si el Track name es el mismo, entonces en una nueva columna añadiría ese dato para esa fila.

SOLUCIONADO CON UN MERGE!!!

In [12]:
#Después de muchos intentos, encontré una forma superfácil de hacerlo
data_top200_merged = pd.merge(data_top200, data_top200_sumastreams, on=['spotify_id'], how='inner')
data_top200_merged

# Comprobaciones
# data_top200_sumastreams[data_top200_sumastreams['spotify_id'] == '7k4t7uLgtOxPwTpFmtJNTY']
# data_top200[data_top200['spotify_id'] == '7k4t7uLgtOxPwTpFmtJNTY']

Position         Track Name         Artist  Streams_x       date  \
0             1               Tusa        KAROL G     446086 2020-03-14   
1             1               Tusa        KAROL G     438199 2020-03-15   
2             1               Tusa        KAROL G     474411 2020-03-16   
3             1               Tusa        KAROL G     495344 2020-03-17   
4             1               Tusa        KAROL G     489980 2020-03-18   
...         ...                ...            ...        ...        ...   
24395       190    Cómo Te Atreves          Morat      28963 2019-05-13   
24396       195  Someone You Loved  Lewis Capaldi      25786 2019-05-12   
24397       187  Someone You Loved  Lewis Capaldi      29291 2019-05-13   
24398       140      Hola Señorita    Maître Gims      36831 2019-05-13   
24399       197    Pa que lo gocen  fernandocosta      27956 2019-05-13   

                   spotify_id  year  Streams_y  
0      7k4t7uLgtOxPwTpFmtJNTY  2020   18848892  
1      7k4t7uLgtOxPwTpFmtJNTY  2020   18848892  
2      7k4t7uLgtOxPwTpFmtJNTY  2020   18848892  
3      7k4t7uLgtOxPwTpFmtJNTY  2020   18848892  
4      7k4t7uLgtOxPwTpFmtJNTY  2020   18848892  
...                       ...   ...        ...  
24395  7M6CFruBrM5x7u0lTMtm6r  2019      54967  
24396  2TIlqbIneP0ZY1O0EzYLlc  2019      55077  
24397  2TIlqbIneP0ZY1O0EzYLlc  2019      55077  
24398  5vLEmh5EolySKTvXsyWSOg  2019      36831  
24399  7BMdZsDXxPH6Db3Ug6A2q2  2019      27956  

[24400 rows x 8 columns]

In [13]:
#Sacar lista de canciones únicas para luego sacar las Features
list_ids = data_top200_merged['spotify_id'].tolist()
len(list_ids)

24400

In [45]:
#Quitar duplicados 
list_ids = list(set(list_ids))
len(list_ids)

test_list = list_ids[0:200]


# Data from API SPOTIFY



Proceso:

- Investigar cómo sacar features en serie con ID de canciones con API de Spotify
- Descargar features de las listas de top200
- Juntar Dataframes de features con el de top200 

In [1]:
# Instalación previa

#!pip install spotipy

In [2]:
#Preparación de Spotipy
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import pandas as pd
import time 

passw = pd.read_csv("pass_spotify.txt", sep = ',', encoding="utf-8")
client_id = passw.columns[0]
client_secret = passw.columns[1]

client_credentials_manager = SpotifyClientCredentials(client_id, client_secret)
sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager)

In [3]:
# Función para sacar las features de canciones que me interesan desde la id
def getTrackFeatures(id):
  meta = sp.track(id)
  features = sp.audio_features(id)

  # meta
  name = meta['name']
  album = meta['album']['name']
  artist = meta['album']['artists'][0]['name']
  release_date = meta['album']['release_date']
  length = meta['duration_ms']
  popularity = meta['popularity']

  # features
  acousticness = features[0]['acousticness']
  danceability = features[0]['danceability']
  energy = features[0]['energy']
  instrumentalness = features[0]['instrumentalness']
  liveness = features[0]['liveness']
  loudness = features[0]['loudness']
  speechiness = features[0]['speechiness']
  valence = features[0]['valence']
  tempo = features[0]['tempo']
  time_signature = features[0]['time_signature']
  id = features[0]['id']

  track = [name, album, artist, release_date, length, popularity, danceability, acousticness, danceability, energy, instrumentalness, liveness, loudness, speechiness, valence, tempo, time_signature, id]
  return track

In [ ]:
tracks = []
for i in range(len(list_ids)):
    # time.sleep(.5)
    track = getTrackFeatures(list_ids[i])
    tracks.append(track)

In [43]:
# Loop para sacar features de todas las canciones de una lista
# OJO, tarda bastante dependendo del número de canciones, por lo que antes hago un testeo
import time
list_toextract = test_list

tracks = []

Time1 = datetime.now()


for i in range(len(list_toextract)):
    time.sleep(.5)
    track = getTrackFeatures(list_toextract[i])
    tracks.append(track)
    
Time2 = datetime.now()


Time1_v = datetime.now()

for i in range(len(list_toextract)):
    # time.sleep(.5)
    track = getTrackFeatures(list_toextract[i])
    tracks.append(track)
    
Time2_v = datetime.now()

print("Tiempo 1:", Time2 -Time1)
print("Tiempo 2:", Time2_v -Time1_v)

1: 0:00:11.418735
2: 0:00:01.355204


In [47]:
# Lo ejecuto con toda la lista al final
import time
list_toextract = list_ids

tracks = []

Time1 = datetime.now()
for i in range(len(list_toextract)):
    track = getTrackFeatures(list_toextract[i])
    tracks.append(track)
Time2 = datetime.now()

print("Tiempo ejecución:", Time2 -Time1)

Tiempo ejecución: 0:02:37.647754


In [51]:
# Meterlo en Dataframe
data_final = pd.DataFrame(tracks, columns = ['name', 'album', 'artist', 'release_date',
                                             'length', 'popularity', 'danceability',
                                             'acousticness', 'danceability', 'energy',
                                             'instrumentalness', 'liveness', 'loudness',
                                             'speechiness', 'valence','tempo', 'time_signature', 'id'])

data_final_selected = data_final [['album','release_date','length', 'popularity', 'danceability',
                                             'acousticness', 'danceability', 'energy',
                                             'instrumentalness', 'liveness', 'loudness',
                                             'speechiness', 'valence','tempo', 'time_signature', 'id']]


# Exportar a CSV

#df.to_csv("spotify.csv", sep = ',')




In [55]:
data_final_selected.head(2)

album release_date  length  popularity  danceability  danceability  \
0      Ellos   2019-03-11  175081           1         0.754         0.754   
1  No Me Ame   2020-04-17  180000          75         0.680         0.680   

   acousticness  danceability  danceability  energy  instrumentalness  \
0        0.1240         0.754         0.754   0.939           0.00281   
1        0.0411         0.680         0.680   0.766           0.00000   

   liveness  loudness  speechiness  valence    tempo  time_signature  \
0     0.293    -1.953       0.0387    0.853  122.006               4   
1     0.116    -4.799       0.0923    0.280  160.095               4   

                       id  
0  1tDUwlIyFFcmfMypYoQ4Ns  
1  2AuWgceaQyfMoI3EGavwGb

In [ ]:
# Unir Dataframes!!!!